components/suggestion/command_provider/app_command_parser.ts

```diff
@@ -43,72 +41,582 @@ export type Store = {
+            raw_command: parsed.command,
         };
-        return payload;
+    }
+
+    // matchBinding finds the closest matching command binding.
+    matchBinding = async (command: string, commandBindings: AppBinding[], autocompleteMode = false): Promise<ParsedCommand> => {
+        const parsed = new ParsedCommand(command);
+        let bindings = commandBindings;
+        let i = 0;
+
+        let done = false;
+        while (!done) {
+            let c = '';
+            if (i < command.length) {
+                c = command[i];
+            }
+
+            switch (Number(parsed.state)) {
+            case ParseState.Start: {
+                if (c !== '/') {
+                    return this.parseError('command must start with a /: ' + i);
+                }
+                if (bindings.length === 0) {
+                    return this.parseError('no command bindings: ' + i);
+                }
+                i++;
+                parsed.incomplete = '';
+                parsed.incompleteStart = i;
+                parsed.state = ParseState.Command;
+                break;
+            }
+
+            case ParseState.Command: {
+                switch (c) {
+                case '': {
+                    if (autocompleteMode) {
+                        // Finish in the Command state, 'incomplete' will have the query string
+                        done = true;
+                    } else {
+                        parsed.state = ParseState.EndCommand;
+                    }
+                    break;
+                }
+                case ' ':
+                case '\t': {
+                    parsed.state = ParseState.EndCommand;
+                    break;
+                }
+                default:
+                    parsed.incomplete += c;
+                    i++;
+                    break;
+                }
+                break;
+            }
+
+            case ParseState.EndCommand: {
+                const binding = bindings.find((b: AppBinding) => b.label === parsed.incomplete);
+                if (!binding) {
+                    // gone as far as we could, this token doesn't match a sub-command.
+                    // return the state from the last matching binding
+                    done = true;
+                    break;
+                }
+                parsed.binding = binding;
+                parsed.location += '/' + binding.label;
+                if (binding.bindings) {
+                    bindings = binding.bindings;
+                } else {
+                    bindings = [];
+                }
+                parsed.state = ParseState.CommandSeparator;
+                break;
+            }
+
+            case ParseState.CommandSeparator: {
+                switch (c) {
+                case '': {
+                    // Reached the end of input.
+                    parsed.incomplete = '';
+                    parsed.incompleteStart = i;
+                    done = true;
```

Comment:
Could we end with a `parsed.state = ParseState.Command;` here (since we are `done` anyway) ? and remove the check for `CommandSeparator` outside this method?